pkg/localstorev2/internal/reserve/reserve.go

```diff
@@ -0,0 +1,358 @@
+	if err != nil {
+		return nil, err
+	}
+
+	rs.size = size
+
+	return rs, nil
+}
+
+func (r *Reserve) Putter(store internal.Storage) storagev2.Putter {
+
+	return storagev2.PutterFunc(func(ctx context.Context, chunk swarm.Chunk) error {
+
+		indexStore := store.IndexStore()
+		chunkStore := store.ChunkStore()
+
+		po := swarm.Proximity(r.baseAddr.Bytes(), chunk.Address().Bytes())
+
+		has, err := indexStore.Has(&batchRadiusItem{
+			Bin:     po,
+			Address: chunk.Address(),
+			BatchID: chunk.Stamp().BatchID(),
+		})
+		if err != nil {
+			return err
+		}
+		if has {
+			return nil
+		}
+
+		switch item, loaded, err := stampindex.LoadOrStore(indexStore, reserveNamespace, chunk); {
+		case err != nil:
+			return fmt.Errorf("load or store stamp index for chunk %v has fail: %w", chunk, err)
+		case loaded && item.ChunkIsImmutable:
+			return errOverwriteOfImmutableBatch
+		case loaded && !item.ChunkIsImmutable:
+			prev := binary.BigEndian.Uint64(item.BatchTimestamp)
+			curr := binary.BigEndian.Uint64(chunk.Stamp().Timestamp())
+			if prev >= curr {
+				return errOverwriteOfNewerBatch
+			}
+			err = stampindex.Store(indexStore, reserveNamespace, chunk)
+			if err != nil {
+				return fmt.Errorf("failed updating stamp index: %w", err)
+			}
+		}
+
+		err = chunkstamp.Store(indexStore, reserveNamespace, chunk)
+		if err != nil {
+			return err
+		}
+
+		binID, err := r.incBinID(indexStore, po)
+		if err != nil {
+			return err
+		}
+
+		err = indexStore.Put(&batchRadiusItem{
+			Bin:     po,
+			Address: chunk.Address(),
+			BatchID: chunk.Stamp().BatchID(),
+			BinID:   binID,
+		})
+		if err != nil {
+			return err
+		}
+
+		err = indexStore.Put(&chunkBinItem{
+			Bin:     po,
+			BinID:   binID,
+			Address: chunk.Address(),
+		})
+		if err != nil {
+			return err
+		}
+
+		return chunkStore.Put(ctx, chunk)
+	})
+}
+
+func (r *Reserve) Has(store storagev2.Store, addr swarm.Address) (bool, error) {
+	_, err := chunkstamp.Load(store, reserveNamespace, addr)
+	if errors.Is(err, storage.ErrNoStampsForChunk) {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (r *Reserve) Get(ctx context.Context, storage internal.Storage, addr swarm.Address) (swarm.Chunk, error) {
+	stamp, err := chunkstamp.Load(storage.IndexStore(), reserveNamespace, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	ch, err := storage.ChunkStore().Get(ctx, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	return ch.WithStamp(stamp), nil
+}
+
+func (r *Reserve) IterateBin(store storage.Store, bin uint8, startBinID uint64, cb func(swarm.Address, uint64) (bool, error)) error {
+	err := store.Iterate(storagev2.Query{
+		Factory:       func() storagev2.Item { return &chunkBinItem{} },
+		Prefix:        binIDToString(bin, startBinID),
+		PrefixAtStart: true,
+	}, func(res storagev2.Result) (bool, error) {
+		item := res.Entry.(*chunkBinItem)
+		if item.Bin > bin {
+			return true, nil
+		}
+
+		stop, err := cb(item.Address, item.BinID)
```

Comment:
I dont think this will work. We will need to store the BatchID in the chunkBinItem. Both these functions need to read the right stamp for the chunk. Also this needs to be in the key as chunks with multiple stamps need to be pullsynced separately so that the peers get all the stamps associated with the chunk.

This is again the same for Sampling. In future if we delimit the sampling based on batch value, we need to consider all different stamps for the chunk.

Reply:
👍